### PR TITLE
Code coverage on errors and schema_helper

### PR DESCRIPTION
**Pull request by dcdarrell9**
### What is the context of this PR?
- Added new functions to check whether there are any duplicate aliases
- Removed unused code 

### How to review 
- Check to see tests work correct
- Check that code was definitely unused

- Still some left to do on errors.py, not in for next two days, didn't want to leave as is.

### CODE REVIEW

File: tests/app/data/test_schema_validation.py
Comment (jonnyshaw89) [inline tests/app/data/test_schema_validation.py]: can we give v and k more logical name?
it names reading the function a lot easier.

File: app/helpers/schema_helper.py
Comment (jonnyshaw89) [inline app/helpers/schema_helper.py]: What now catches this exception?

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

File: app/helpers/schema_helper.py
Comment (collisdigital) [inline app/helpers/schema_helper.py]: maybe output the `answer['alias']` as part of the message?

File: tests/app/data/test_schema_validation.py
Comment (collisdigital) [inline tests/app/data/test_schema_validation.py]: Do we need to ignore these for this checker?

File: tests/app/data/test_schema_validation.py
Comment (dcdarrell9) [reply tests/app/data/test_schema_validation.py]: What would be the best way to name them? The _parse_id_values uses the same so can change that as well as.

File: tests/app/data/test_schema_validation.py
Comment (collisdigital) [inline tests/app/data/test_schema_validation.py]: This function can be merged with the one above by pulling the `'alias'` into a parameter to it...

**Review by collisdigital — CHANGES_REQUESTED** [review]
Nice to see some more code removed :+1:

File: app/helpers/schema_helper.py
Comment (collisdigital) [reply app/helpers/schema_helper.py]: It's an assert, i.e. should never happen, so nothing will catch it

File: tests/app/data/test_schema_validation.py
Comment (dcdarrell9) [reply tests/app/data/test_schema_validation.py]: How would you do this?
The way I just changed it to (_parse_values)... makes the test fail for both alias and id if there is only one duplicate

File: tests/app/data/test_schema_validation.py
Comment (LJBabbage) [inline tests/app/data/test_schema_validation.py]: They are the 'key' and the 'value'

File: app/helpers/schema_helper.py
Comment (LJBabbage) [reply app/helpers/schema_helper.py]: No maybe, definitely, otherwise debug is a pain

File: tests/app/data/test_schema_validation.py
Comment (LJBabbage) [inline tests/app/data/test_schema_validation.py]: The function needs a name change or the content needs reducing. If its purely dealing with alias' why are we worrying about 171 down.

File: tests/app/data/test_schema_validation.py
Comment (dcdarrell9) [inline tests/app/data/test_schema_validation.py]: Issue with this is it won't check alias if there is a failure in id.

Comment (samiwel) [reply tests/app/data/test_schema_validation.py]: I think checking for duplicate Ids and duplicate aliases should be two separate tests. Just makes it clearer when one of them fails, you could re-use much of the logic here by extracting a method and passing 'id' or 'alias' as a parameter.

File: app/helpers/schema_helper.py
Comment (samiwel) [inline app/helpers/schema_helper.py:138]: I'd question the need for an assert here at all. It's not a bad thing per se, and you may have a good reason for having it here, I don't know. But I think having a test in `test_schema_validation.py` and catching it at build time is sufficient. Since we're not dealing with the AssertionError at runtime anyway. Your call though.

**Review by samiwel — CHANGES_REQUESTED** [review]
Looks good Darrel, I've added a few minor comments. But feel free to disagree with me if you want. 👍

File: app/helpers/schema_helper.py
Comment (collisdigital) [reply app/helpers/schema_helper.py:138]: in theory asserts are removed from production code (you have to turn on the Python optimise flags) so it's a developer aid in case anything ever goes wrong (some code changes break it for example or the schema validator stopped working)

File: tests/app/data/test_schema_validation.py
Comment (collisdigital) [inline tests/app/data/test_schema_validation.py]: you could remove all this `if` logic if you reset `unique_id` at the start of the for loop `for key in special_key:` e.g.

```python
for key in special_key:
    unique_items = []
    ....
```

File: app/helpers/schema_helper.py
Comment (samiwel) [reply app/helpers/schema_helper.py:138]: Yeah, that's fair enough, I don't have strong views on this. Just feels a bit odd checking for duplicates in the same place where we're retrieving the values, but it's an additional safeguard I guess.

**Review by collisdigital — APPROVED** [review]
Looks good; @samiwel do you approve as well? If so @dcdarrell9 can get this rebased, squashed and we can merge.

**Comment by collisdigital** [comment]
@jonnyshaw89 happy with this PR now?

File: app/helpers/schema_helper.py
Comment (jonnyshaw89) [reply app/helpers/schema_helper.py:138]: @collisdigital do we have the Python optimise flag currently set in production?

Comment (collisdigital) [reply app/helpers/schema_helper.py:138]: Not that I know of; I think when we checked we would need to change the EB instance internally to turn it on, there be dragons!

**Review by samiwel — APPROVED** [review]

File: app/helpers/schema_helper.py
Comment (samiwel) [reply app/helpers/schema_helper.py:138]: Codacy flags this up as an issue, if we're agreeing leaving the assert in is okay, is there some way of configuring codacy not to flag it up?

Comment (jonnyshaw89) [reply app/helpers/schema_helper.py:138]: I think we either need to use asserts and enable the Python optimise flag or not use asserts.

Comment (ajmaddaford) [reply app/helpers/schema_helper.py:138]: We have used asserts elsewhere in the code, so I've created a card for reviewing whether we should or not. For now I've disabled the Codacy rule.

**Review by jonnyshaw89 — APPROVED** [review]

**Pull request merged**
